### PR TITLE
Add must_use annotations to Result::is_ok and is_err

### PR DESCRIPTION
**Pull request by alex**
Discussed in #59610

### CODE REVIEW

**Comment by alex** [comment]
This seems to have caught several cases in some tests, e.g. https://github.com/rust-lang/rust/blob/master/src/librustc_data_structures/owning_ref/mod.rs#L1474

I think that's missing an `assert!()`, does that sound right?

**Comment by scottmcm** [comment]
If this is worth it on `is_err`, wouldn't it also need to be on `is_some` and such as well?

**Comment by alexcrichton** [comment]
@alex want to fix the issues found on CI? 

It seems reasonable to have similar treatment on `Option`, and I would prefer that we not get too overzealous and try to match all `is_*` functions automatically

**Comment by alex** [comment]
@alexcrichton Yes, I'd like to! I could use some feedback on the question I asked WRT these tests: https://github.com/rust-lang/rust/blob/master/src/librustc_data_structures/owning_ref/mod.rs#L1467-L1485

`try_map1` and `try_map2` look identical to me, except one calls `is_ok` and the other `is_err` and I'm honestly unclear on what either of them is trying to assert.

**Comment by alexcrichton** [comment]
I don't know myself (didn't write those tests) but it seems harmless to either throw in `assert!` or `assert!(!...)` depending on what works

**Comment by alex** [comment]
👍  will update to get these tests passing this evening.

**Comment by alexcrichton** [comment]
@alex did you want to add similar treatment to `is_some` and `is_none` in this PR?

**Comment by alex** [comment]
If you'd prefer it as a part of this PR, I'm happy to.

**Comment by alexcrichton** [comment]
Yes since they're so similar in purpose let's go ahead and add them in this PR

**Comment by alex** [comment]
Another failing test, again I'm not sure I fully understand what this one is trying to test. I think `let _ = ...;` is the correct fix?

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

File: src/libcore/option.rs [hunk 1]
```diff
@@ -178,6 +178,7 @@ impl<T> Option<T> {
     /// ```
     ///
     /// [`Some`]: #variant.Some
+    #[must_use]
```
Comment (RalfJung) [inline src/libcore/option.rs]: Might be a good idea to add a message here explaining that this function has no side-effects, or so?

**Comment by Centril** [comment]
@bors retry

Yielding in favor of r0llup this is included in.

File: src/libcore/option.rs (hunk 1 above)
Comment (scottmcm) [reply src/libcore/option.rs]: Agreed!  Maybe with a pointer to `unwrap` or `expect` if people meant it to be an assertion?

Comment (czipperz) [reply src/libcore/option.rs]: Did this ever get resolved?  If not someone should open a pr for it.

Comment (RalfJung) [reply src/libcore/option.rs]: Nope, there's still no message. Can you open an issue or a PR?

Comment (czipperz) [reply src/libcore/option.rs]: I would but I honestly have no clue what to put for the message.

Comment (scottmcm) [reply src/libcore/option.rs]: @czipperz Well, inspired by https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect, here's a proposal for the text:

- `Result::is_ok`: if you intended to assert that this matches `Ok(_)`, consider `.unwrap()` instead
- `Result::is_err`: if you intended to assert that this matches `Err(_)`, consider `.unwrap_err()` instead
- `Option::is_some`: if you intended to assert that this matches `Some(_)`, consider `.unwrap()` instead
- `Option::is_none`: if you intended to assert that this matches `None`, consider ```.or_else(|| panic!("called `Option::unwrap()` on a `None` value"))``` instead

(That last one makes me think of https://internals.rust-lang.org/t/adding-option-expect-none/10481?u=scottmcm...)

Comment (czipperz) [reply src/libcore/option.rs]: I'll open a PR based on those messages tonight if someone else doesn't do it first.

**Pull request merged**